### PR TITLE
fetch-failed-logs: use GraphQL API

### PR DESCRIPTION
**Pull request by jonchang**
This should vastly reduce the number of API calls needed to locate a failed workflow associated with a pull request.

Ideally we'd do _everything_ in graphql but most GitHub Actions related APIs haven't been implemented yet, so we need to do a mix of the two.

This also splits the fetching logic into separate functions, with an eye towards permitting passing pull request or workflow URLs as the argument, rather than just formulae. (For example, we currently can't get failed logs for new formulae).

This also changes the logic of the file matcher - previously, it would only match the file that was changed on the tip of the head branch, meaning that pull requests that modified multiple formulae wouldn't match correctly. This change gets up to 100 files that have changed (the graphQL query limit).

### CODE REVIEW

**Comment by dawidd6** [comment]
Those GraphQL queries give me creeps anytime I see them, but LGTM.

File: cmd/fetch-failed-logs.rb [hunk 1]
```diff
@@ -64,6 +64,61 @@ def get_failed_lines(file)
...
+                    }
+                  }
+                  headRefName
+                }
+              }
+            }
+          }
+        }
+      EOS
+    }
```
Comment (maxim-belkin) [inline cmd/fetch-failed-logs.rb:108]: I suggest making this part a little bit less intimidating...

```ruby
data = { 
  query:<<~EOS.gsub("\n", ' ').squeeze(' ').strip
  {
    repository(name: "#{name}", owner: "#{owner}") {
      pullRequests(first: 100, states: OPEN, orderBy: {field: CREATED_AT, direction: DESC}) {
          edges { node { number files(first: 100) { nodes { path } } headRefName } }
      }
    }
  }
  EOS
}
```

Comment (dawidd6) [reply cmd/fetch-failed-logs.rb:108]: That makes it less readable IMO.

Comment (maxim-belkin) [reply cmd/fetch-failed-logs.rb:108]: how is it less readable?

Comment (maxim-belkin) [reply cmd/fetch-failed-logs.rb:108]: Well, all I can say is that my version preserves monitor's vertical real estate (which is more valuable than horizontal), what, in turn, helps identify matching curly brackets.

**Pull request closed without merging**